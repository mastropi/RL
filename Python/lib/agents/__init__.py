# -*- coding: utf-8 -*-
"""
Created on Wed Apr  4 18:44:23 2020

@author: Daniel Mastropietro
@description: Definition of classes defining agents.
Each agent is created with a policy and a learner class and implements the following methods:
    - getPolicy(): returns the Policy object of the agent. Note that the policy can be a dictionary of policies.
    - getLearner(): returns the Learner object of the agent. Note that the learner can be a dictionary of learners.
    - getLastAction(): returns the last action taken by the agent.
    - setLastAction(): sets the last action taken by the agent.
"""

# If below we use `import *` we get a warning generated by Spyder: "unable to detect undefined names"
# This is due to the fact that it doesn't know which modules are being imported, because they are not listed.
# It seems we can disable the warnings by adding the comment #NOQA but it doesn't work.
# Ref: https://stackoverflow.com/questions/50439309/spyder-ide-complaining-about-unable-to-detect-undefined-names
# where they actually also mention that it is a common pattern to do this import * in an __init__.py file!
#from .policies.random_walks import PolRandomWalkDiscrete
#from .learners.td_lambda import LeaTDLambda

class GenericAgent:
    """
    Class that holds the information of the policy, the learner, and the action associated to an agent

    These objects are very generic, they can be defined differently depending on the subclass.
    For instance the AgeQueue subclass defines policies and learners as dictionaries of policies and learners,
    respectively, because e.g. there are two policies at least that could be defined on a queue: the job acceptance
    policy, and the assignment policy of a job to a server in the queue.
    Likewise, the learner could contain two types of learners, the value functions learner and the policy learner.
    """

    # TODO: (2021/10/18) Probably we should define the attributes as `policies` and `learners` because the agent may have different policies and differen learners
    # (Ex: policy to accept a new incoming job, policy to assign a job to a server; learner for the state-value function, learner for the policy, etc.)
    def __init__(self, policy, learner):
        self.policy = policy
        self.learner = learner
        self.action = None

    def act(self, env, policy_type):
        raise NotImplementedError

    #----- GETTERS -----#
    def getPolicy(self):
        return self.policy

    def getLearner(self):
        return self.learner

    def getLastAction(self):
        "Returns the last action performed by the agent"
        return self.action

    #----- SETTERS -----#
    def setLastAction(self, action):
        self.action = action
