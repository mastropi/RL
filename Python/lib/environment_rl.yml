# DM-2023/05/30
# To be used to create a conda virtual environment by running:
#   conda env create -f environment_rl.yml
name: rl_toulouse
channels:
  - conda-forge
  - defaults
dependencies:
  - gym==0.12.1
  - numpy==1.18.3
  - pandas==0.22.0
  - pytest==3.3.2
