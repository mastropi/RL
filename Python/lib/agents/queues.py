# -*- coding: utf-8 -*-
"""
Created on Thu Oct 14 13:43:28 2021

@author: Daniel Mastropietro
@description: Definition of agents acting on queue environments
"""

from enum import Enum, unique

if __name__ == "__main__":
    import runpy
    runpy.run_path("../../../setup.py")

from Python.lib.agents import GenericAgent
from Python.lib.agents.learners import LearnerTypes
from Python.lib.agents.policies import PolicyTypes
from Python.lib.environments.queues import Actions, ActionTypes


class AgeQueue(GenericAgent):
    """
    Agent that acts on a queue environment

    Arguments:
    env: Queue environment
        The queue environment on which the agent acts is NOT stored in the class because we may want
        the agent to act on *different* environments having the same or similar characteristics
        (e.g. in Fleming-Viot learning, where there are N copies of the same environment on which the same type of agent
        interacts --by e.g. accepting or rejecting incoming jobs).

        The attribute is left as part of the constructor because, conceptually, agents should receive the environment on
        which they act, so that we homogenize the signature of all agents we define, by passing the environment,
        the policies, and the learners.

        The queue environment on which the agent acts is passed as argument to the act() method defined in the class.

    policies: dict
        Dictionary of policies followed by the agent, with the following elements:
        - PolicyTypes.ACCEPT: Policy object defining the acceptance policy of an incoming job. The acceptance policy
        may depend on the class of the incoming job. If so, the policy object defined by this dictionary key should be
        a list of policies with as many elements as job classes accepted by the queue environment on which the policy
        is applied.
        - PolicyTypes.ASSIGN: Policy object defining the assignment policy of an incoming job class to a server.
            The assignment policy can be None, in which case the assignment of job classes to servers follows a uniform probability
            over all the servers in the queue.

    learners: dict
        Dictionary of learners, with the following elements:
        'V': GenericLearner used by the agent to learn the state value function.
        'Q': GenericLearner used by the agent to learn the action value function.
    """
    def __init__(self, env, policies: dict, learners: dict, debug=False):
        super().__init__(policies, learners)
        self.debug = debug

        # Check whether the acceptance policy depends on the arriving job class or is the same for all job classes
        # We assume the policy depends on the job class when the `policies` parameter is a list of policies with
        # as many elements as job classes on which the policy is applied.
        # NOTE that we cannot call the self.getAcceptancePolicies() method defined below because that method USES
        # the value of the attribute we define here self.acceptance_policy_depends_on_jobclass whose value we set now!
        if isinstance(self.getPolicy()[PolicyTypes.ACCEPT], list):
            self.acceptance_policy_depends_on_jobclass = True
        else:
            self.acceptance_policy_depends_on_jobclass = False

    def act(self, env, policy_type):
        """
        The agent performs an action that is chosen by the policy of the given type.
        The action performed is stored in the object and can be retrieved with the getLastAction() method.

        Arguments:
        env: Queue environment (normally defined in environments/queues.py)
            The queue environment on which the agent acts.
            It should have the following methods defined:
            - getActions(): gets the possible actions accepted by the environment.
            - getBufferSize(): gets the current queue length, i.e. the sum of the occupancy levels of each job class
                received by the queue system defined in the environment.
            - getCapacity(): gets the capacity of the queue system defined in the environment.
            - getJobClass(): gets the job class associated to the current state of the environment. If not None
                this is the class of the arriving job at the time an action is required by the agent.
            - getLastAction(): get the last action taken by the agent on the environment.
            - getNumServers(): gets the number of servers in the queue system defined in the environment.
            - getQueueState(): gets the queue state of the queue system defined in the environment.
            - getState(): gets the current state of the environment.
            - step(): the method that takes the action chosen by the agent on the environment.
            NOTE: This object is updated after the action takes place with the result of taking the action of the given
            policy_type chosen by the agent.

        policy_type: PolicyTypes
            Type of policy on which the agent acts on the queue, one of:
            - PolicyType.ACCEPT which is used to decide whether an incoming job is accepted to the queue
            - PolicyType.ASSIGN which is used to decide to which server a the class of an accepted incoming job is assigned

        Return: tuple
        Tuple containing the following elements:
        - action: the action taken by the agent on the given policy
        - observation: the next state on which the queue transitions to after the action taken
        - reward: the reward received by the agent after taking the action and transitioning to the next state
        - info: dictionary with relevant additional information
        """
        if not self.isValidPolicyType(policy_type):
            raise ValueError("Invalid policy type to act on. Possible types are: {} ({})" \
                             .format(PolicyTypes, policy_type))

        job_class = env.getJobClass()

        if policy_type == PolicyTypes.ACCEPT:
            # Choose the action to take and store it in the agent object
            action_type = ActionTypes.ACCEPT_REJECT
            if self.acceptance_policy_depends_on_jobclass:
                # This is most likely the case when the environment is a loss network with multi-class jobs
                # Note that the action of the agent is REJECT when the system is at FULL CAPACITY (regardless of the class of the job that arrives)
                # If the system is NOT at full capacity, the accept/reject action is chosen based on the class of the arriving job
                # and the occupancy level of that class in the system, which are inputs for the acceptance policy.
                assert job_class < len(env.getQueueState()), "The job class ({}) must be an index whose value is less than the queue state length ({})".format(job_class, len(env.getQueueState()))
                buffer_size_is_smaller_than_queue_capacity = env.getBufferSize() < env.getCapacity()
                # The action taken is the minimum between two possible actions:
                # - accepting/rejecting the job based on the system being at full capacity or not.
                # - accepting/rejecting the job based on the policy for the class of the arriving job.
                # The fact that we use the minimum assumes that the possible actions defined in the Actions enum
                # associates REJECT to a smaller index than ACCEPT, and in addition the way to retrieve the chosen action
                # requires that REJECT is associated to index 0 and ACCEPT is associated to index 1 (because the chosen
                # action is derived from the above boolean expression which is converted to integer on the next line).
                assert Actions.ACCEPT.value > Actions.REJECT.value, "The REJECT action ({}) must have a lower index than the ACCEPT action ({}) in the Actions enum".format(Actions.REJECT, Actions.ACCEPT)
                assert Actions.REJECT.value == 0 and Actions.ACCEPT.value == 1, "The REJECT action ({}) must have index value 0 and the ACCEPT action ({}) must have index value 1 in the Actions enum".format(Actions.REJECT, Actions.ACCEPT)
                action = min( env.getActions()(int(buffer_size_is_smaller_than_queue_capacity)),
                              self.getAcceptancePolicies()[job_class].choose_action((env.getQueueState()[job_class], job_class)) )
                    ## IMPORTANT: The state that is passed to the policy's choose_action() method MUST be the state of the queue environment on which the policy acts
                    ## So for instance, for queue environments defined in environments/queues.py (such as EnvQueueSingleBufferWithJobClasses or EnvQueueLossNetworkWithJobClasses)
                    ## the state is a tuple containing the queue's state and the arriving job class.
                    ## The message here is that we should NOT simply pass the queue's state as parameter to choose_action() but also
                    ## the information of the arriving job class, as the queue's state is extacted from such tuple (queue_state, job_class)
                    ## by the getPolicyForAction() method of the parameterized policy that is called by the choose_action method called above.

                    ## NOTE also that, as queue_state we pass the occupancy level of the arriving job class because this is the piece of information of the state
                    ## that defines the acceptance or rejection of the job... i.e. the occupancy level of the other job classes do not play any role in this acceptance policy.
                assert isinstance(action, Actions)
            else:
                # This is most likely the case when the environment is a single- or multi-server system with single queue (a.k.a. buffer)
                action = self.getAcceptancePolicies()[0].choose_action(env.getState())
            self.setLastAction(action)
        elif policy_type == PolicyTypes.ASSIGN:
            # Choose the action to take and store it in the agent object
            # In an ASSIGN policy type the action is the server number to which the arriving job is assigned
            if env.getNumServers() > 1:
                assert self.getAssignmentPolicy() is not None, "The assignment policy must be defined in multi-server systems"
                servers = range(env.getNumServers())
                # TODO: (2021/10/21) Homogenize the signature of choose_action() among policies so that they all receive one parameter: the current state (but I am not sure whether this is possible)
                action = self.getAssignmentPolicy().choose_action(job_class, servers)
            else:
                # The action is choose the only server in the system, server #0
                action = 0
            action_type = ActionTypes.ASSIGN
            self.setLastAction(action)

        # Take the action, observe the next state, and get the reward received when transitioning to that state
        observation, reward, info = env.step(action, action_type)

        # Update the action taken by the agent after the environment performed the step triggered by the action chosen by the agent above
        # Note that the action can change after the environment received it from the agent and tried to make a step consistent with that action because,
        # that action was actually not possible to be performed given the environment conditions.
        # This is the case for instance when a policy (that depends on the job class --i.e. self.acceptance_policy_depends_on_jobclass = True)
        # accepts a job based solely on the server occupancy of THAT ARRIVING JOB CLASS, but doesn't know anything about the system's occupancy
        # (when the occupancy of all job classes in the system are taken into account), and the system's occupancy happens to be at full capacity!
        # In such case, the job is actually rejected (by the environment, not by the policy). So here we update the action actually taken by the agent
        # (rejection instead of its initial approval)
        # Impacting the agent's action in this way is important, because o.w. an assertion written in the manage_job_arrival() function
        # in simulators/queues.py would fail (because the assertion asserts that the reward is 0.0 when the action is ACCEPT, which would NOT be the case
        # should the agent's action not be updated as done here.
        # Note however that the last action taken by the agent (stored in the agent object) is NOT updated... so we can still know
        # what action the agent took by retrieving the agent's last action with agent.getLastAction().
        action = env.getLastAction()

        return action, observation, reward, info

    def learn(self, t, state, reward, done, info):
        # TODO: (2021/10/20) Should the agent learn or should we leave this implementation to the simulator that learns? (see e.g. simulators.py/SimulatorQueue)
        pass

    #------ GETTERS ------#
    def getAcceptancePolicies(self):
        "Returns a *list* with the acceptance policies defined in the agent, either when only one policy is defined or when multiple policies are defined"
        policies_accept = self.getPolicy()[PolicyTypes.ACCEPT]
        if self.acceptance_policy_depends_on_jobclass:
            # The agent defines several acceptance policies, e.g. one per arriving job class
            # (which of course could be just one job class, in which case still the policy is stored as a list --of one policy)
            # => Return all the acceptance policies defined in the agent
            return policies_accept
        else:
            # There is only one policy defined in the agent
            # => Convert the acceptance policy to a list containing the acceptance policy so that callers to this method
            # have to deal only with one case, namely the case of *multiple* policies stored in a list.
            return [policies_accept]

    def getAcceptancePolicyThresholds(self):
        "Returns the thresholds (theta parameters) of the acceptance policies defined in the agent. Each acceptance policy is expected to have a SCALAR theta parameter, that's why we use the word 'threshold' in the method's name."
        return [pol.getThetaParameter() for pol in self.getAcceptancePolicies()]

    def getAcceptancePolicyIntegerThresholds(self):
        thresholds = self.getAcceptancePolicyThresholds()
        policies_accept = self.getAcceptancePolicies()
        assert isinstance(thresholds, list)
        assert isinstance(policies_accept, list)
        return [policies_accept[p].getDeterministicBlockingValueFromTheta(theta) for p, theta in enumerate(thresholds)]

    def getAcceptancePolicyDependenceOnJobClass(self):
        return self.acceptance_policy_depends_on_jobclass

    def getAssignmentPolicy(self):
        return self.getPolicy()[PolicyTypes.ASSIGN]

    def getLearnerV(self):
        return self.getLearner()[LearnerTypes.V]

    def getLearnerQ(self):
        return self.getLearner()[LearnerTypes.Q]

    def getLearnerP(self):
        return self.getLearner()[LearnerTypes.P]

    #------ SETTERS ------#
    def setAcceptancePolicyThresholds(self, thresholds):
        policies_accept = self.getAcceptancePolicies()
        # Set the threshold of each acceptance policy
        if thresholds is None or len(thresholds) != len(policies_accept):
            raise ValueError("The `thresholds` parameter must not be None and its length must be equal to the number of policies defined in the agent")
        for i, policy in enumerate(policies_accept):
            policy.setThetaParameter(thresholds[i])

    def setAssignmentPolicy(self, policy_assign):
        # Note that we cannot invoke the getAssignmentPolicy() method defined above
        # because it is not possible to assign a value to a function call (this is the error we get when we try to use that method on the LHS of the equality)
        self.getPolicy()[PolicyTypes.ASSIGN] = policy_assign

    #------ CHECKERS ------#
    def isValidPolicyType(self, policy_type):
        return policy_type in PolicyTypes
