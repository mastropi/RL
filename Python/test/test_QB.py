# -*- coding: utf-8 -*-
"""
Created on Thu Jun  4 19:16:02 2020

@author: Daniel Mastropietro
@description: Estimators are defined for problems of interest in Reinforcement Learning.
"""

import runpy
runpy.run_path('../../setup.py')

import numpy as np
import unittest
from unittest_data_provider import data_provider
import matplotlib.pyplot as plt

import Python.lib.queues as queues
import Python.lib.estimators as estimators
from Python.lib.queues import Event
from Python.lib.estimators import FinalizeType, RenderType

#from importlib import reload
#reload(estimators)
#from Python.lib.estimators import EstimatorQueueBlockingFlemingViot

#import test_utils


class Test_QB_Particles(unittest.TestCase):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.log = False
        self.rate_birth = 1.2
        self.rate_death = 1.4
        self.nservers = 1
        self.capacity = 3
        self.queue = queues.QueueMM(self.rate_birth, self.rate_death, self.nservers, self.capacity)

        self.plotFlag = True

    def no_tests_on_one_queue(self):
        print("\nRunning test " + self.id())
        # Cannot remove non-existing job
        self.queue.reset()
        self.queue.remove()
        assert self.queue.size == 0

        # Add a job increases the queue size by one in non-limit situations
        self.queue.reset()
        self.queue.add()
        assert self.queue.size == 1

        # Cannot add a job in a full queue
        self.queue.reset()
        for i in range(self.capacity):
            self.queue.add()
        assert self.queue.size == self.capacity
        self.queue.add()
        assert self.queue.size == self.capacity

    def no_test_simulation_on_one_particle(self):
        print("\nRunning test " + self.id())
        niter = 100
        est = estimators.EstimatorQueueBlockingFlemingViot(1, niter, self.queue,
                                                           reactivate=True)
        
        # The state udpate process is correctly done. Let aNET = "array with Next Event Times",
        # then the following assertions hold.
        est.reset()
        print("The test runs on a simulation with {} iterations" \
              "\n(no pre-specified seed, so that different test runs produce even more tests!)".format(niter))
        for it in range(niter):
            aNET = est.get_times_next_events(0)
            aNET_prev = aNET.copy()
            est.update_state(it+1)
            #print("aNET_prev: {}".format(aNET_prev))
            #print("time last change: {}".format(est.get_time_last_event(0)))
            
            assert len(aNET) == 2, \
                    "There are always two elements in the aNET array containing the next event times"
            
            assert est.get_time_last_event(0) == np.min(aNET_prev), \
                    "The applied event has the smallest time among those stored in the previous value of aNET"
    
            type_event_applied = est.get_type_last_event(0)
            type_event_not_applied = 1 - type_event_applied 
            assert  aNET[type_event_not_applied] > 0 and aNET[type_event_not_applied] == \
                    aNET_prev[type_event_not_applied] - aNET_prev[type_event_applied], \
                    "The updated time-to-event of the event type NOT applied is equal to the previous time-to-event " + \
                    "minus its previous time-to-event value, and is POSITIVE"
    
            time_next_birth = aNET[queues.BIRTH]
            time_next_death = aNET[queues.DEATH]
            assert est._order_times_next_events[0] == [queues.BIRTH, queues.DEATH] and time_next_birth < time_next_death \
                or est._order_times_next_events[0] == [queues.DEATH, queues.BIRTH] and time_next_birth > time_next_death, \
                    "The list that stores the order of next event times reflects the true order"

    def tests_on_n_particles(self,  reactivate=True,
                                    finalize_type=FinalizeType.REMOVE_CENSORED,
                                    nparticles=5,
                                    niter=25,
                                    seed=1713): 
        print("\nRunning test " + self.id())
        #nparticles = 30
        #niter = 200
        #reactivate = True
        #finalize_type = FinalizeType.ABSORB_CENSORED
        #finalize_type = FinalizeType.REMOVE_CENSORED
        #seed = 1713
        est = estimators.EstimatorQueueBlockingFlemingViot(nparticles, niter, self.queue,
                                                           reactivate=reactivate,
                                                           finalize_type=finalize_type,
                                                           render_type=RenderType.GRAPH,
                                                           seed=seed, log=False)
        print("Simulation setup:")
        print(est.setup())
        
        est.reset()
        for it in range(niter):
            est.update_state(it+1)
            N1 = n_active_particles = est.get_number_active_particles()
            if np.mod(it, int(niter/10)) == 0:
                print("Iteration {} of {}... ({} active particles)".format(it+1, niter, N1))

            if self.log:
                print("------ END OF ITER {} of {} ------".format(it+1, niter))

            # The list storing the times when particles became active is sorted increasingly
            particles, activation_times = est.get_all_activation_times()
            if self.log:
                with np.printoptions(precision=3):
                    print("activation times: {}".format(np.array(activation_times)))
            assert sorted(activation_times) == activation_times, \
                    "The absolute times of activation are sorted: {}".format(np.array(activation_times))
            assert len(activation_times) >= N1, \
                    "The number of activation times ({}) is at least equal to the number of active particles (N1 = {})" \
                    .format(len(activation_times), N1)

            particles, absorption_times = est.get_all_absorption_times()
            if self.log:
                with np.printoptions(precision=3):
                    print("absorption times: {}".format(np.array(absorption_times)))
            assert sorted(absorption_times) == absorption_times, \
                    "The (relative) absorption times are sorted: {}".format(np.array(absorption_times))            

            # The list storing the time segments where statistics are computed is sorted
            survival_time_segments = est.get_survival_time_segments()
            if self.log:
                with np.printoptions(precision=3):
                    print("time segments: {}".format(np.array(survival_time_segments)))
            assert sorted(survival_time_segments) == survival_time_segments, \
                    "The time segments are sorted: {}".format(np.array(survival_time_segments))
            assert survival_time_segments[0] == 0.0, \
                    "The first element of the time segments list ({}) is equal to 0" \
                    .format(survival_time_segments[0])
            #assert survival_time_segments[1:] == absorption_times, \
            #        "The ending time segments ({}) are equal to the absorption times ({})" \
            #        .format(survival_time_segments, absorption_times)

            counts_particles_alive_by_elapsed_time = est.get_counts_particles_alive_by_elapsed_time()
            if self.log:
                print("Counts by segment: {}".format(counts_particles_alive_by_elapsed_time))
            assert np.all( counts_particles_alive_by_elapsed_time == np.arange(len(survival_time_segments)-1, -1, -1) ), \
                    "Since the time segments are defined by jumps in the count of active particles" \
                    " they should decrease linearly to 0 starting at {} ({})" \
                    .format(len(survival_time_segments)-1, counts_particles_alive_by_elapsed_time)
            if self.log:
                print("------ END OF ITER {} of {} ------".format(it+1, niter))

        print("\n\n****** SIMULATION SUMMARY ({} iterations) ******".format(niter))

        #************************************************************************
        # GRAPHICAL RENDER
        if True:
            print("\nSIMULATION RENDER (before treatment of censored values):")
            print(est.render())
        #************************************************************************


        #************************************************************************
        # FINALIZE THE SIMULATION PROCESS BY DEALING WITH ACTIVE PARTICLES (censoring)
        est.finalize()
        print("\n\n****** ESTIMATION PROCESS ({} iterations) ******".format(niter))
        #************************************************************************


        #************************************************************************
        # GRAPHICAL RENDER
        if True:
            print("\nSIMULATION RENDER (after treatment of censored values):")
            print(est.render())
        #************************************************************************



        # TODO: (2020/06/14) Move the call to compute_counts() inside estimate_proba_blocking()
        print("\nESTIMATIONS *** METHOD 2: FROM observed TIMES ***:")
        est.compute_counts()
        df_proba_survival_and_blocking_conditional = est.estimate_proba_survival_and_blocking_conditional()
        with np.printoptions(precision=3, suppress=True):
            print("Estimated probabilities by time:")
            print(df_proba_survival_and_blocking_conditional)

        if False and est.finalize_type != FinalizeType.REMOVE_CENSORED:
            # Only compare probabilities when censored times are NOT removed
            print("\nEqual survivals and blocking probs?")
            df_comparison = (df_proba_survival_and_blocking_conditional_BF == df_proba_survival_and_blocking_conditional)
            equal = np.min(np.min(df_comparison))
            print(equal)
            if not equal:
                print(df_comparison)

        if self.log:
            print("Total blocking time by particle:")
            for p in range(est.N):
                df_blocking_periods = est.get_blocking_periods(p)
                print("\n\nParticle {}".format(p))
                print("Block periods:")
                print(df_blocking_periods)
                
                df_survival_periods = est.get_survival_periods(p)
                print("\nSurvival periods:")
                print(df_survival_periods)
        
        
                total_blocking_time = est.get_total_blocking_time(p)
                total_survival_time = est.get_total_survival_time(p)
                print("")
                print("Total blocking time = {:.3f}".format(total_blocking_time))
                print("Total survival time = {:.3f}".format(total_survival_time))
                print("% blocking time = {:.1f}%".format(total_blocking_time / total_survival_time * 100))
                print("\n")

        print("Total blocking time for ALL particles:")
        all_total_blocking_time = est.get_all_total_blocking_time()
        all_total_survival_time = est.get_all_total_survival_time()
        print("")
        print("Total blocking time = {:.3f}".format(all_total_blocking_time))
        print("Total survival time = {:.3f}".format(all_total_survival_time))
        print("% blocking time = {:.1f}%".format(all_total_blocking_time / all_total_survival_time * 100))
        print("\n")

        proba_blocking = est.estimate_proba_blocking()
        print("\nBlocking probability estimate: {:.1f}%".format(proba_blocking*100))
        print("Rough estimate (rho^K) (rho={:.1f}, K={}): {:.1f}%" \
              .format(self.rate_birth / self.rate_death, self.capacity, (self.rate_birth / self.rate_death)**self.capacity*100))

        print("Simulation setup:")
        print(est.setup())

        self.plot_results(df_proba_survival_and_blocking_conditional, 
                          reactivate,
                          finalize_type,
                          nparticles,
                          niter)

        return  df_proba_survival_and_blocking_conditional, \
                (reactivate, finalize_type, nparticles, niter)

    def plot_results(self, df_proba_survival_and_blocking_conditional, *args):
        reactivate = args[0]
        finalize_type = args[1]
        nparticles = args[2]
        niter = args[3]
        print(reactivate)
        print(finalize_type.name)
        print(nparticles)
        print(niter)

        plt.figure()
        color1 = 'blue'
        color2 = 'red'
        #y2max = 0.05
        y2max = 1.1*np.max(df_proba_survival_and_blocking_conditional['P(BLOCK / T>t,s=1)'])
        plt.step(df_proba_survival_and_blocking_conditional['t'], df_proba_survival_and_blocking_conditional['P(T>t / s=1)'],
                 'b-', where='post')
        ax = plt.gca()
        ax.spines['left'].set_color(color1)
        ax.tick_params(axis='y', color=color1)
        ax.yaxis.label.set_color(color1)
        ax2 = ax.twinx()
        ax2.step(df_proba_survival_and_blocking_conditional['t'], df_proba_survival_and_blocking_conditional['P(BLOCK / T>t,s=1)'],
                 'r-', where='post')
        ax2.set_ylim(0, y2max)
        ax.spines['right'].set_color(color2)
        ax2.tick_params(axis='y', color=color2)
        ax2.yaxis.label.set_color(color2)
        plt.sca(ax)
        ax.legend(['P(T>t / s=1)'], loc='upper left')
        ax2.legend(['P(BLOCK / T>t,s=1)'], loc='upper right')
        plt.title("K={}, rate(B)={:.1f}, rate(D)={:.1f}, rho={:.1f}, reactivate={}, final={}, N={}, #iter={}" \
                  .format(self.capacity,
                          self.queue.rates[Event.BIRTH.value],
                          self.queue.rates[Event.DEATH.value],
                          self.queue.rates[Event.BIRTH.value] / self.queue.rates[Event.DEATH.value],
                          reactivate, finalize_type.name[0:3],
                          nparticles, niter
                          ))
        ax.title.set_fontsize(9)


if __name__ == "__main__":
    unittest.main()

else:
    # Lines to execute "by hand" (i.e. at the Python prompt)
    test = Test_QB_Particles()

    nparticles = 30
    niter = 50
    seed = 1717

    #####
    reactivate = False
    finalize_type = FinalizeType.ABSORB_CENSORED
    df_resultsFA, paramsFA = test.tests_on_n_particles(reactivate,
                                                    finalize_type,
                                                    nparticles,
                                                    niter,
                                                    seed)
    test.plot_results(df_resultsFA, *paramsFA)

    #####
    reactivate = True
    finalize_type = FinalizeType.ABSORB_CENSORED
    df_resultsTA, paramsTA = test.tests_on_n_particles(reactivate,
                                                    finalize_type,
                                                    nparticles,
                                                    niter,
                                                    seed)
    test.plot_results(df_resultsTA, *paramsTA)

    #####
    reactivate = False
    finalize_type = FinalizeType.REMOVE_CENSORED
    df_resultsFR, paramsFR = test.tests_on_n_particles(reactivate,
                                                    finalize_type,
                                                    nparticles,
                                                    niter,
                                                    seed)
    test.plot_results(df_resultsFR, *paramsFR)

    #####
    reactivate = True
    finalize_type = FinalizeType.REMOVE_CENSORED
    df_resultsTR, paramsTR = test.tests_on_n_particles(reactivate,
                                                    finalize_type,
                                                    nparticles,
                                                    niter,
                                                    seed)
    test.plot_results(df_resultsTR, *paramsTR)
