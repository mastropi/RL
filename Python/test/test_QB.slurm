#!/bin/sh
#SBATCH --job-name=test_QB
#SBATCH --mail-type=END
#SBATCH --mail-user=mastropi@uwalumni.com
#SBATCH --output=test_QB.slurm.out
date
python3.6 test_QB.py 3 J 15 9
date
