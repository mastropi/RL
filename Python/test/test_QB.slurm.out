Fri Aug 20 17:18:40 CEST 2021
/home/dmastrop/projects/Python/lib/estimators.py:3420: FutureWarning: from_items is deprecated. Please use DataFrame.from_dict(dict(items), ...) instead. DataFrame.from_dict(OrderedDict(items)) may be used to preserve the key order.
  ('Blocking Time Estimate', self.blocking_time_estimate)])
/home/dmastrop/projects/Python/lib/estimators.py:3306: FutureWarning: from_items is deprecated. Please use DataFrame.from_dict(dict(items), ...) instead. DataFrame.from_dict(OrderedDict(items)) may be used to preserve the key order.
  return pd.DataFrame.from_items([('t', self.sk), ('P(T>t / s=1)', proba_surv_by_t)])
test_QB.py:822: FutureWarning: from_items is deprecated. Please use DataFrame.from_dict(dict(items), ...) instead. DataFrame.from_dict(OrderedDict(items)) may be used to preserve the key order.
  ('ratio_mc_fv_events', [dict_stats_mc.get('nevents') is None and np.nan or dict_stats_mc.get('nevents') / dict_stats_fv['nevents']]),
/usr/lib64/python3.6/site-packages/matplotlib/pyplot.py:514: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  max_open_warning, RuntimeWarning)
/home/dmastrop/projects/Python/lib/utils/plotting.py:363: UserWarning: More than one subplot was requested but all 'x' and 'y' variables to plot are the same.
  warn("More than one subplot was requested but all 'x' and 'y' variables to plot are the same.")
Directory:
/home/dmastrop/projects
has been prepended to the module search path.
User arguments: ['test_QB.py', '3', 'J', '15', '9']
Parsed user arguments: ['test_QB.py', '3', 'J', '15', '9', 1, 'save']

Execution parameters:
nservers=3
var_analysis=J
replications=15
tests2run=[9]
run_mc=False
save_results=True
Log file '../../RL-002-QueueBlocking/logs/estimates_vs_J_20210820_171842.log' has been open for output.
Started at: 2021-08-20 17:18:42
Ended at: 2021-09-04 19:21:32
Execution time: 21722.8 min, 362.0 hours
Sat Sep  4 19:21:40 CEST 2021
